### ISSUE TITLE
Word-wrap character is invisible or uses the wrong hyphen with some fonts.

### ISSUE THREAD

**Issue by mbandsmer** [koreader/koreader#9074]
* KOReader version: koreader-android-x86-v2022.03.1.apk
* Device: Bluestacks android emulator on Windows

#### Issue
With some fonts, e.g. Castoro available from Google Fonts, the hyphen character used for word-wrapping at the end of a line does not print.  

With other fonts, e.g. Cochineal, the wrong hyphen (an M-dash?) is used for the word-wrap character at the end of a line.

Screenshots are attached.  Other e-readers (e.g. FBReader) do not have this problem with these fonts.

#### Steps to reproduce
1. Install Castoro font from Google and/or Cochineal font from the CTAN archive.
2. Open an ebook and select Castoro or Cochineal as the active font.  Observe the hyphens used for word-wrapping at the end of lines.
![Castoro-word-wrap-NoHyphen](https://user-images.githubusercontent.com/65569771/167228932-e811cbdb-00ae-4419-8648-46f2206f0573.png)
![Cochineal-word-wrap-hyphen](https://user-images.githubusercontent.com/65569771/167228955-f6ddae74-7ded-45e0-8b5f-bd4577313f33.png)

**Comment by poire-z** [comment]
Previous related issues: #8290 #8407.
Dunno how the other readers do, but these looks like a font (or FreeType) bug.
(We're not really fighting with old fonts, fonts are getting better and fixed, and we trust the libraries like FreeType and Harfbuzz to do the best they can with what the font gives. It's easier to try new or other fonts, than fighting with the font rendering stack...)

(We have `Font > Settings > Generate font test document` to quickly see how all your fonts do. When I'm testing new fonts, I can quickly see if they have bugs with KOReader, no matter whether it's the font fault or KOReader's, and throw them out.)

**Comment by khaledhosny** [comment]
From the above issues, it seems that you are using U+00AD to render the hyphen, us this right?

If so, this is the wrong code point to show a visible hyphen. Soft Hyphen is a control character that indicates where hyphenation should happen, but it does not have a visible rendering of it's own and Unicode-complient shaping engines like HarfBuzz will show it blank by default regardless of what glyph the font has for it.

You should draw the hyphenation using U+002D HYPHEN-MINUS or U+2010 HYPHEN (some applications will try U+2010 first then fallback to U+002D).

**Comment by poire-z** [comment]
> it seems that you are using U+00AD to render the hyphen, us this right?

Yes, [crengine](https://github.com/koreader/crengine) has always been using U+00AD for this, I had never questionned it :) as it has been working quite allright with most fonts (except the rare ones I assumed were buggy, often some aged fonts).
crengine does the hyphen width handling and glyph drawing by itself: Harfbuzz doesn't see it (and if there is any real U+00AD soft hyphen in the unicode input stream, HB indeed ignores it - and when not using HB (this is toggable on our side), we ignore it ourself).

U+2010 feels like the right char/glyph to check for and use. Dunno about U+002D - I have some vague memory I tried using it, and the width/look I got with some fonts was worse than using U+00AD.
We don't have (yet) a fallback mecanism to check for and pick various codepoints in the main font, but we pick U+00AD in fallback fonts if not present in previous fonts.
I may experiment with this to see how it does (although it might not be fun to decide, if it turns out it solves issues with some fonts, but give other issues with fonts that worked fine...)

@khaledhosny : as you happen to be visiting us :), may I request your thoughts about another issue (didn't dare bothering you HarfBuzz people with it, but you're probably the best to answer it):
https://github.com/koreader/koreader/issues/5359#issuecomment-1003589894 and followup posts (it's a bit long, sorry): how to truncate arabic words (and other scripts you may know about), and the questions of the italic slant direction with arabic (please answer in #5359 if you do).
Thank you.

**Comment by poire-z** [comment]
> I may experiment with this to see how it does (although it might not be fun to decide, if it turns out it solves issues with some fonts, but give other issues with fonts that worked fine...)

Did some tests and comparaisons.

U+002D = ASCII minus / hyphen (so, a bit ambiguous for being hybrid), the classic `-`
U+00AD = Soft-hyphen, what we have been using for ages
U+2010 = Unicode hyphen, non ambiguous codepoint

It turns out that:
- most good/recent/popular reading fonts have the same glyph for U+002D, U+00AD, U+2010
- for those who don't:
  - U+002D is always the best (except for one single font, a chinese one I got from I don't know where: SimSun/NSimSun)
  - U+00AD may be wrong, too large, or absent
  - U+2010 is often present, but too large (with blank side bearings, the size of the hyphen, so actually 3x the size of the hyphen - unless it's a bug/edge case in crengine, because it looks similarly bad with many fonts...)
  - for Noto Nashk Arabic, Noto Nastaliq Urdu: U+002D is good, U+00AD is too large, and U+2010 is too low (like an underscore `_` - could this have some purpose with the Arabic script ?)

So, it looks like we can move to use U+002D (and just ignore U+2010 as it might often be worse), and I guess ALL fonts must have U+002D.

(I thought U+002D , being also the minus sign, could have some fancy vertical align to fit digits, but it looks like it's just fine to be used with words.)
I didn't find anything in crengine old commit history 2008-2012 that could explain why they went with U+00AD.

**Comment by khaledhosny** [comment]
U+2010 is what the font designer designs as a hyphen, if it is different from U+002D I assume it is intentional. In the case of the Arabic fonts above, it is probably to account for the Uyghur use (the only language using Arabic script that allows for hyphenation) where the hyphen sets on the base line but disconnected from the letter before it.

U+002D has dual purpose so designers often compromise on its design, and since it is in ASCII might try to stick to more ”expected” proportions.

My suggestion would to use U+2010 when supported by the font, and U+002D otherwise. If doing coverage check is not possible, then U+002D is the safer choice.

**Comment by poire-z** [comment]
Thanks, this explains the reasons for what I was observing:

> U+2010 is often present, but too large (with blank side bearings, the size of the hyphen, so actually 3x the size of the hyphen - unless it's a bug/edge case in crengine, because it looks similarly bad with many fonts...)

So, it's not a bug, but as U+2010 is often absent, it was picked from our first fallback font, Noto Sans CJK, whose _what the font designer designs as a hyphen_ is actually a em-square (like all ideographs) glyph containing a small hyphen in its middle :)
![image](https://user-images.githubusercontent.com/24273478/167366275-364480a6-ab83-4832-9a1b-f9e5a238cb9b.png) (and so is the apostrophe...)

The Arabic hyphen : ![image](https://user-images.githubusercontent.com/24273478/167366512-69616d38-812e-48ee-bd59-93186969d90a.png)

> My suggestion would to use U+2010 when supported by the font, and U+002D otherwise. If doing coverage check is not possible, then U+002D is the safer choice.

I think we will still go with U+002D, just because of the existence of the CJK and the Arabic hyphens, as we explicitely don't hyphenate words in fullwidth ASCII chars and RTL words. It's possible a user has a CJK or Arabic as its main font, for the main CJK or Arabic text, and some embedded English/western language text could still be candidate for hyphenation: we don't want to get it hyphenated with the CJK or Arabic hyphen, which would make it look odd as on the pictures above.

**Comment by mbandsmer** [comment]
Thanks for the detailed analysis.  According to this research on the soft-hyphen (a little dated but possibly still relevant)
https://jkorpela.fi/shy.html#rendering
different web browsers render the soft-hyphen differently: Internet Explorer always used U+002D, and Firefox/Chrome use U+2010 if available.

I think always using U+002D is a logical / safe approach based on your findings, but to accommodate the Uyghur Arabic case, perhaps the U+2010 alternative could be made available as an advanced hyphenation option in the user interface?

**Comment by poire-z** [comment]
> but to accommodate the Uyghur Arabic case, perhaps the U+2010 alternative could be made available as an advanced hyphenation option in the user interface?

No, not needed - as explained above, a Uyghur word, being Arabic so RTL, won't go thru the hyphenation code, even if it has soft-hyphens (that could/would be used in a LTR word). If we are to ever support that, we'll have to add some dedicated code to draw a hyphen on the left of a LTR word, where we could take some specific action to use/find a U+2010 in an Arabic font among our fallback fonts.

Also, as also explained above, the way we use fallback fonts is non-academic (different than how web browsers do it - and our hyphenation is also different), but fast - and using U+2010 could end up having it grabbed from a CJK or Arabic font among the fallbacks, and would look odd/ugly as we only hyphenate non-CJK and non-RTL/Arabic.
So, really safer to use U+002D (too bad if the font designer has made some micro nice adjustments to have U+2010 different from the ambiguous U+002D...)

Please confirm with tomorrow's nightly that the issue with your 2 fonts is solved, thanks you.

**Comment by mbandsmer** [comment]
I can confirm that the hyphenation is fixed with Castoro and Cochineal fonts in the nightly build
